### PR TITLE
Add development guide

### LINKED ISSUE #484: Provide docs on how to build & test the codebase
Provide docs on how to build the codebase so that people can get started.

Should include how to :
- Build our solution
- Run unit tests
- Build Docker image

Location is still TBD - Either README or [CONTRIBUTING](https://github.com/tomkerkhove/promitor/blob/master/.github/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by adamconnelly**
I've added a new document with information about how to build and test Promitor, along with an overview of the repository layout.

Fixes #484

### CODE REVIEW

**Comment by adamconnelly** [comment]
@tomkerkhove I'm not sure if this is quite what you're looking for, but I figured it's a starting point at least.

I've left a TODO about the `deploy` directory because I'm not quite sure what it's for, so if you can give me an outline of it I can update that.

File: development-guide.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+Next, edit [src/metric-config.yaml](src/metric-config.yaml), configure some metrics
+and set the following keys:
+
+- `azureMetadata.tenantId` - your Azure tenant Id.
+- `azureMetadata.subscriptionId` - your Azure subscription Id.
+- `azureMetadata.resourceGroupName` - the default resource group to use if none
+  is specified for a metric.
+
+Finally, run the Promitor.Docker project.
```
Comment (tomkerkhove) [inline development-guide.md]: Maybe also add a note that they have to add the scrapers they want to use?

File: development-guide.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+```shell
+dotnet test
+```
+
+If you want to perform TDD, you can use the `dotnet watch` command to watch for
+changes and re-run the tests:
+
+```shell
+dotnet watch test
```
Comment (tomkerkhove) [inline development-guide.md]: TIL - Didn't know about it :D

File: development-guide.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+The promitor repository is made up of a number of different directories:
+
+- `/` - contains documentation about the repository including the main README file
+  and some configuration files for various tools used for development.
+- `/.github` - contains github specific files including issue and PR templates.
+- `/.vscode` - contains shared configuration files for [VS Code](https://code.visualstudio.com/).
+- `/build` - contains the configuration files for Promitor's CI process.
+- `/charts` - contains Promitor's [Helm](https://helm.sh/) chart.
+- `/deploy` - TODO.
```
Comment (tomkerkhove) [inline development-guide.md]: I can help you with that:

`contains the automation that is being used to manage Promitor, such as the automated updates concerning new Docker images in a pull request.`

**Comment by tomkerkhove** [comment]
One thing we could add is how to build the Docker image outside of VS.

1. Navigate to /src
2. Run `docker build . --file .\Promitor.Scraper.Host\Dockerfile --tag promitor --no-cache` should work

File: development-guide.md (hunk 1 above)
Comment (adamconnelly) [reply development-guide.md]: Sure - I've reworded this slightly and added a link to the configuration docs on the website.

File: development-guide.md (hunk 2 above)
Comment (adamconnelly) [reply development-guide.md]: I really like it - I just wish that .NET build times were faster so there wasn't so much waiting between changes!

**Comment by adamconnelly** [comment]
> One thing we could add is how to build the Docker image outside of VS.
> 
> 1. Navigate to /src
> 2. Run `docker build . --file .\Promitor.Scraper.Host\Dockerfile --tag promitor --no-cache` should work

Ok - I've added that.

What we might want to consider for future is making it easier to build and deploy the chart locally. For example, if we tag the image with same same tag referenced in the Helm chart, you can use your local image instead of having to pull it from the docker repo while doing local dev.

So we could create a values.local.yaml file that sets the tag to `tomkerkhove/promitor-agent-scraper:local`, and use that in the docker build command, then specify `--values values.local.yaml` when running the helm upgrade / install command. We do something pretty similar at work for running containers in a local cluster for development.

File: development-guide.md (hunk 2 above)
Comment (tomkerkhove) [reply development-guide.md]: Great to know, you learn something every day!

**Comment by tomkerkhove** [comment]
> > One thing we could add is how to build the Docker image outside of VS.
> > 
> > 1. Navigate to /src
> > 2. Run `docker build . --file .\Promitor.Scraper.Host\Dockerfile --tag promitor --no-cache` should work
> 
> Ok - I've added that.
> 
> What we might want to consider for future is making it easier to build and deploy the chart locally. For example, if we tag the image with same same tag referenced in the Helm chart, you can use your local image instead of having to pull it from the docker repo while doing local dev.
> 
> So we could create a values.local.yaml file that sets the tag to `tomkerkhove/promitor-agent-scraper:local`, and use that in the docker build command, then specify `--values values.local.yaml` when running the helm upgrade / install command. We do something pretty similar at work for running containers in a local cluster for development.

Would you mind opening an issue for this please? It's not fully clear to me what you are suggesting but don't want to block this PR on it!

**Comment by tomkerkhove** [comment]
Thanks!

**Comment by adamconnelly** [comment]
> Would you mind opening an issue for this please? It's not fully clear to me what you are suggesting but don't want to block this PR on it!

No problem - it's really not a big deal, and it's not really related to this PR. It'll be easier for me to create an example to explain it.

**Pull request merged**